r/src/bpcells-cpp/matrixUtils/Pseudobulk.cpp

```diff
@@ -0,0 +1,112 @@
+// Copyright 2024 BPCells contributors
+// 
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// https://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or https://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#include "Pseudobulk.h"
+namespace BPCells {
+
+// Take a matrix and do a pseudobulk aggregation on the matrix.
+// Note that when calculating a more complex statistic, the simpler ones are also calculated.  The order of complexity is NonZeros < Sum < Mean < Variance.
+// Args:
+// - mat: Matrix to compute variance from.
+// - cell_groups: Mapping of columns to groups.
+// - method: method to use for calculating pseudobulk stats.
+// - transpose: Whether the matrix is transposed. If false, groups columns and calculates stats per-row. If true, groups rows and calculates stats per-column.
+// Returns:
+// - PseudobulkStats struct containing each of the requested pseudobulk aggregations, in the shape of (num_features, num_groups).
+template <typename T>
+PseudobulkStats pseudobulk_matrix(std::unique_ptr<MatrixLoader<T>> &&mat,
+                                  const std::vector<uint32_t>& cell_groups,
+                                  PseudobulkStatsMethod method,
+                                  bool transpose,
+                                  std::atomic<bool> *user_interrupt) {
+    MatrixIterator<T> it(std::move(mat));
+    std::vector<double> group_count_vec;
+    struct PseudobulkStats res;
+    for (auto& cell_group : cell_groups) {
+        if (cell_group >= group_count_vec.size()) {
+            group_count_vec.resize(cell_group + 1);
+        }
+        group_count_vec[cell_group]++;
+    }
+    Eigen::Map<Eigen::Array<double, 1, Eigen::Dynamic>> group_count(group_count_vec.data(), group_count_vec.size());
+    // If transposed, also keep matrix transposed with groups as rows for cache-friendliness
+    uint32_t num_rows = transpose ? group_count_vec.size() : it.rows();
+    uint32_t num_cols = transpose ? it.cols() : group_count_vec.size();
+    if ((method & PseudobulkStatsMethod::NonZeros) == PseudobulkStatsMethod::NonZeros) res.non_zeros = Eigen::ArrayXXd::Zero(num_rows, num_cols);
```

Comment:
To allow for the bitwise checks, I just ended up comparing equality to the enum itself,  is this what you had in mind?